core/src/main/java/org/lflang/analyses/evm/Instruction.java

```diff
@@ -0,0 +1,40 @@
+package org.lflang.analyses.evm;
+
+public abstract class Instruction {
```

Comment:
A suggestion to get rid of the boilerplate: use Kotlin :)

```kotlin
enum class OpCode {
    ADV,
    ADV2,
    BIT
}

// Sealed classes are abstract by default, and the compiler
// knows all their subtypes to allow for exhaustivity checking
// when pattern-matching.
sealed class EvmInstruction(val opcode: OpCode)

// Data classes get a nice toString generated
data class InstructionADV(
    val reactor: ReactorInstance,
    val nextTimeValue: TimeValue
) : EvmInstruction(OpCode.ADV2)

// Since this instruction has no fields it can be a singleton
object InstructionBIT : EvmInstruction(OpCode.BIT)

// ...
```

This entire package could fit in 100 lines in a single file :)

You can refer to Kotlin classes from Java as if they were regular Java classes. But if you want to stick to Java I would recommend making the supertype sealed (`public abstract sealed class Instruction`). ([doc](https://docs.oracle.com/en/java/javase/17/language/sealed-classes-and-interfaces.html#GUID-0C709461-CC33-419A-82BF-61461336E65F))

Also I would suggest giving this `Instruction` class a more specific name like `EvmInstruction` for clarity

Reply:
Thanks for the mini Kotlin tutorial here, @oowekyala ! Yes, Kotlin looks quite interesting. I will try to reimplement the instructions using your solution.